lib/mongo-milestone-db.js

```diff
@@ -0,0 +1,227 @@
+        return collection.updateOne(query, snapshot, options);
+      });
+  }
+
+  _getMilestoneSnapshot(collectionName, id, version) {
+    return this._collection(collectionName)
+      .then((collection) => {
+        const query = {id: id};
+        if (version != null) {
+          query.v = {$lte: version};
+        }
+
+        return collection
+          .find(query)
+          .project({_id: 0})
+          .sort({v: -1})
+          .limit(1)
+          .next();
+      })
+      .then((snapshot) => {
+        if (snapshot) {
+          snapshot.m = snapshot.m == null ? null : snapshot.m;
+        }
+
+        return snapshot || undefined;
+      });
+  }
+
+  _close() {
+    return this._db()
+      .then((db) => {
+        this._mongoPromise = null;
+        return db.close();
+      });
+  }
+
+  _db() {
+    if (!this._mongoPromise) {
+      return Promise.reject(new MongoClosedError());
+    }
+
+    return this._mongoPromise;
+  }
+
+  _collection(collectionName) {
+    let name;
+    let collection;
+
+    return this._db()
+      .then((db) => {
+        name = MongoMilestoneDB._milestoneCollectionName(collectionName);
+        collection = db.collection(name);
+
+        if (this._shouldCreateIndex(name)) {
+          // WARNING: Creating indexes automatically like this is quite dangerous in
+          // production if we are starting with a lot of data and no indexes
+          // already. If new indexes were added or definition of these indexes were
+          // changed, users upgrading this module could unsuspectingly lock up their
+          // databases. If indexes are created as the first ops are added to a
+          // collection this won't be a problem, but this is a dangerous mechanism.
+          // Perhaps we should only warn instead of creating the indexes, especially
+          // when there is a lot of data in the collection.
+          return collection.createIndex({id: 1, v: 1}, {background: true, unique: true});
+        }
+
+        return Promise.resolve();
+      })
+      .then(() => {
+        if (!this._disableIndexCreation) {
+          this._milestoneIndexes.add(name);
+        }
+
+        return collection;
+      });
+  }
+
+  _shouldCreateIndex(milestoneCollectionName) {
+    // Given the potential problems with creating indexes on the fly, it might
+    // be preferable to disable automatic creation
+    return !this._disableIndexCreation
+      && !this._milestoneIndexes.has(milestoneCollectionName);
+  }
+
+  static _milestoneCollectionName(collectionName) {
+    return `m_${ collectionName }`;
+  }
+
+  static _connect(mongo, options) {
+    if (typeof mongo === 'function') {
+      return new Promise((resolve, reject) => {
+        mongo((error, db) => {
+          error ? reject(error) : resolve(db);
+        });
+      });
+    }
+
+    return mongodb.connect(mongo, options);
+  }
+
+  static _shallowClone(object) {
+    if (typeof object !== 'object') return object;
+
+    const out = {};
```

Comment:
This can be written as:
```
return Object.assign({}, object)
```

[`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) copies enumerable own properties from the second argument (then the third arg, fourth arg, etc. if present) onto the first object. Otherwise, if doing a manual loop, a `object.hasOwnProperty()` check would be needed to avoid copying over enumerable prototype properties on `object`.

Reply:
Ha! I somehow completely forgot that I was in ES6 🤦🏻‍♂️